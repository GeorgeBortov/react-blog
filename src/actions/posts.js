import database from '../firebase/firebase';

// ADD_POST
export const addPost = (post) => ({
    type: 'ADD_POST',
    post
});

export const startAddPost = (postData = {}) => {
    return (dispatch, getState) => {
        const authorID = getState().auth.uid;
        const {
            title = '',
            body = '',
            createdAt = 0,
            authorName = '',
        } = postData;
        const post = { title, body, createdAt, authorName, authorID }
        
        return database.ref(`posts`).push(post).then((ref) => {
            dispatch(addPost({
                id: ref.key,
                ...post
            }));
        });
    }
};

// REMOVE_EXPENSE
export const removePost = ({ id } = {}) => ({
    type: 'REMOVE_POST',
    id
});

export const startRemovePost = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`posts/${id}`).remove().then(() => {
            dispatch(removePost({ id }));
        });
    }
};

// EDIT_EXPENSE
export const editPost = (id, updates) => ({
    type: 'EDIT_POST',
    id,
    updates
});

export const startEditPost = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`posts/${id}`).update(updates).then(() => {
            dispatch(editPost(id, updates));
        });
    }
};

// SET_POSTS
export const setPosts = (posts) => ({
    type: 'SET_POSTS',
    posts
});

export const startSetPosts = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        // return database.ref(`users/${uid}/posts`)
        return database.ref(`posts`)
        .once('value')
        .then((snapshot) => {
            const posts = [];

            snapshot.forEach((childSnapshot) => {
                posts.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            })
            dispatch(setPosts(posts));
        });
    }
};