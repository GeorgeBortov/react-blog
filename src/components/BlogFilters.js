import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export class BlogFilters extends React.Component {
    state = {
        
    };
    render() {
        return (
            <div className="page-header">
                <div className="content-container">
                    <form>
                        <div className="input-group">
                            <div className="input-group__item">
                                <input
                                    type="text"
                                    className="text-input"
                                    placeholder="Search posts by title"
                                    // value={this.props.filters.text}
                                    // onChange={this.onTextChange}
                                />
                            </div>
                            <div className="input-group__item">
                                <select
                                    className="select"
                                    // value={this.props.filters.sortBy}
                                    // onChange={this.onSortChange}
                                >
                                    <option value="date">Sort by date</option>
                                    <option value="cur_user">Select only my posts</option>
                                </select>
                            </div>
                        </div>
                    </form>
                    <Link className="button" to="/create">Add Post</Link>
                </div>
            </div>
        )
    }
}



export default connect(undefined, undefined)(BlogFilters);