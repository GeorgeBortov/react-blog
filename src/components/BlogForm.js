import React from 'react';
import { connect } from 'react-redux';
import JoditEditor from "jodit-react";
import moment from 'moment';

export class BlogForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            title: props.post ? props.post.title : '',
            body: props.post ? props.post.body : '',
            createdAt: moment(),
            authorName: props.userName ? props.userName : (props.post.authorName ? props.post.authorName : '' ),
            error: ''
        }
    }
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({ title }));
    };
    onAuthorChange = (e) => {
        const authorName = e.target.value;
        this.setState(() => ({ authorName }));
    };
    onBodyChange = (value) => {
        const body = value;
        this.setState(() => ({ body }));
    };
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.title || !this.state.body) {
            // 'Please provide description and amount'
            this.setState(() => ({ error: 'Please provide title and post body' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                title: this.state.title,
                body: this.state.body,
                createdAt: this.state.createdAt.valueOf(),
                authorName: this.state.authorName
            });
        }
    };
    config = {
		readonly: false // all options from https://xdsoft.net/jodit/doc/
	}
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="Post title"
                    autoFocus
                    className="text-input form--title"
                    value={this.state.title}
                    onChange={this.onTitleChange}
                />
                <JoditEditor
                    editorRef={this.setRef}
                    value={this.state.body}
                    config={this.config}
                    onChange={this.onBodyChange}
                />
                {
                    !this.props.userName &&
                    <input
                        type="text"
                        placeholder="Author"
                        className="text-input form--author"
                        value={this.state.authorName}
                        onChange={this.onAuthorChange}
                    />
                }
                
                <div>
                    <button className="button">Save Post</button>
                </div>
            </form>
        )
    }
}

const mapStateToProps = (state) => ({
    userName: state.auth.name || ''
});

export default connect(mapStateToProps, undefined)(BlogForm);