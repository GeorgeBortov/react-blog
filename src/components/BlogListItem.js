import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import moment from 'moment';
import trimText from '../selectors/trim-text';
export const BlogListItem = ({ id, title, trimText, authorName, createdAt, uid, authorID, }) => (
    <div className="list-item">
        <div>
            <h3 className="list-item__title">{title}</h3>
            <div className="list-item__text">
                {trimText}
            </div>
        </div>
        <div className="list-item__data">
            <span className="list-item__author">Posted by {authorName !== '' ? authorName : 'Anonymous'}</span>
            <span className="list-item__time">{moment(createdAt).format('MMMM Do YYYY')} at {moment(createdAt).format('h:mm')}</span>
        </div>
        <div className="list-item__buttons">
            <Link className="button" to={`/read/${id}`}>Read</Link>
            {   
                uid === authorID &&
                    <Link className="button" to={`/edit/${id}`}>Edit</Link>
                
            }
            
        </div>
    </div>
    
);

const mapStateToProps = (state, ownProps) => ({
    uid: state.auth.uid,
    trimText: trimText(ownProps.body, 500)
});

const ConnectedBlogListItem = connect(mapStateToProps)(BlogListItem);

export default ConnectedBlogListItem;