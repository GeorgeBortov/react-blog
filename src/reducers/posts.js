// Posts Reducer

const postsReducerDefaultState = {
    postsArr: [],
    comentsArr: [],
    postTotalNumb: 0
};

export default (state = postsReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_POST':
            return {
                ...state,
                postsArr: [
                    ...state.postsArr,
                    action.post
                ]
            };
        case 'ADD_COMMENT':
            return {
                ...state,
                comentsArr: [
                    ...state.comentsArr,
                    action.comment
                ]
                // postsArr: state.postsArr.comments.concat(action.comment)
                
            };
        case 'SET_COMMENTS':
            return {
                ...state,
                comentsArr: action.comments, 
            };
        case 'REMOVE_POST':
            return {
                ...state,
                postsArr: state.postsArr.filter(({ id }) => id !== action.id)
            };
        case 'EDIT_POST':
            return {
                ...state,
                postsArr: state.postsArr.map((post) => {
                    if (post.id === action.id) {
                        return {
                            ...post,
                            ...action.updates
                        }
                    } else {
                        return post;
                    }
                })
            };
        case 'SET_POSTS':
            return {
                ...state,
                postsArr: action.posts,
                postTotalNumb: action.postsNumb
            };
        default:
            return state;
    }
};
