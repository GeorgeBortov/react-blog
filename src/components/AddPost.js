import React from 'react';
import { connect } from 'react-redux';
import BlogForm from './BlogForm';
import { startAddPost } from '../actions/posts';

export class AddPost extends React.Component {
    onSubmit = (post) => {
        this.props.startAddPost(post);
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add post</h1>
                    </div>
                </div>
                <div className="content-container">
                    <BlogForm
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddPost: (post) => dispatch(startAddPost(post))
});

export default connect(undefined, mapDispatchToProps)(AddPost);