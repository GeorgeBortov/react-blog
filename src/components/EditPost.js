import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import BlogForm from './BlogForm';
import ConfirmModal from './ConfirmModal';
import { startEditPost, startRemovePost } from '../actions/posts';
import { openModal, closeModal } from '../actions/modal';

export class EditPost extends React.Component {
    onSubmit = (post) => {
        this.props.startEditPost(this.props.post.id, post);
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.startRemovePost({ id: this.props.post.id });
        this.props.closeModal();
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Post</h1>
                    </div>
                </div>
                {
                    this.props.post ? (
                        this.props.uid === this.props.post.authorID ?
                        <div className="content-container">
                            <BlogForm
                                post={this.props.post}
                                onSubmit={this.onSubmit}
                            />
                            <button className="button button--secondary" onClick={this.props.openModal}>Remove Expense</button>
                            <ConfirmModal
                                isOpen={this.props.modalStatus}
                                onRequestClose={this.props.closeModal}
                                postTitle={this.props.post.title}
                                onRemove={this.onRemove}
                            />
                        </div> 
                        :
                        <div className="content-container">
                            <p className="message">
                                You don't allowed to edit this post
                            </p>
                            <Link className="button" to={'/dashboard'}>Back to Dashboard</Link>
                        </div>
                    ) : (
                            <div className="content-container">
                                <p className="message">
                                    This post does not exist
                                </p>
                                <Link className="button" to={'/dashboard'}>Back to Dashboard</Link>
                            </div>
                        )
                }
            </div>
        );
    }
};

const mapStateToProps = (state, props) => ({
    post: state.posts.find((post) => post.id === props.match.params.id),
    modalStatus: state.modal.modalIsOpen,
    uid: state.auth.uid

});

const mapDispatchToProps = (dispatch, props) => ({
    startEditPost: (id, post) => dispatch(startEditPost(id, post)),
    startRemovePost: (data) => dispatch(startRemovePost(data)),
    openModal: () => dispatch(openModal()),
    closeModal: () => dispatch(closeModal()),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPost);