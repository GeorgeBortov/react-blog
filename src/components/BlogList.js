import React from 'react';
import { connect } from 'react-redux';
import BlogListItem from './BlogListItem';


export const BlogList = (props) => {
    return (
        <div className="content-container">
            <div className="list-body">
                {   
                    props.posts.length === 0 ? (
                        <div className="list-item list-item--message">
                            <span>No posts</span>
                        </div>
                    ) : (
                        props.posts.map((post) => {
                            return <BlogListItem key={post.id} {...post} />
                        })
                    )
                }
            </div>
        </div>
    )
};

const mapStateToProps = (state) => ({
    posts: state.posts
});

const ConnectedBlogList = connect(mapStateToProps)(BlogList);

export default ConnectedBlogList;