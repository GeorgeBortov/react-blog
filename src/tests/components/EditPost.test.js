import React from 'react';
import { shallow } from 'enzyme';
import { EditPost } from '../../components/EditPost';
import posts from '../fixtures/posts';


let startEditPost, startRemovePost, history, openModal, closeModal, wrapper;

beforeEach(() => {
    startEditPost = jest.fn();
    startRemovePost = jest.fn();
    openModal = jest.fn();
    closeModal = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditPost
        startEditPost={startEditPost}
        startRemovePost={startRemovePost}
        closeModal={closeModal}
        openModal={openModal}
        history={history}
        post={posts[1]}
        uid="thisismytestuid"
    />);
});

test('should render EditPost correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should render error if post doesn\'t exist', () => {
    const wrapper = shallow(<EditPost
        startEditPost={startEditPost}
        startRemovePost={startRemovePost}
        closeModal={closeModal}
        openModal={openModal}
        history={history}
        uid="thisismytestuid"
    />);
    expect(wrapper).toMatchSnapshot();
});

test('should render error EditPost if uid is not macthed authorID', () => {
    const wrapper = shallow(<EditPost
        startEditPost={startEditPost}
        startRemovePost={startRemovePost}
        closeModal={closeModal}
        openModal={openModal}
        history={history}
        post={posts[1]}
        uid="otheruid"
    />);
    expect(wrapper).toMatchSnapshot();
});

test('should handle startEditPost', () => {
    wrapper.find('Connect(BlogForm)').prop('onSubmit')(posts[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditPost).toHaveBeenLastCalledWith(posts[1].id, posts[1]);
});

test('should handle startRemovePost', () => {
    wrapper.find('ConfirmModal').prop('onRemove')();
    expect(startRemovePost).toHaveBeenLastCalledWith({
        id: posts[1].id
    });
    expect(closeModal).toHaveBeenCalled();
    expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should handle openModal', () => {
    wrapper.find('.button.button--secondary').simulate('click');
    expect(openModal).toHaveBeenCalled();
});
test('should handle closeModal', () => {
    wrapper.find('ConfirmModal').prop('onRequestClose')();
    expect(closeModal).toHaveBeenCalled();
});