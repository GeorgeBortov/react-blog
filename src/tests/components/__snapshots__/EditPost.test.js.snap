// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditPost correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Post
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <Connect(BlogForm)
      onSubmit={[Function]}
      post={
        Object {
          "authorID": "thisismytestuid",
          "authorName": "John Doe",
          "body": "test post body 2",
          "createdAt": -345600000,
          "id": "2",
          "title": "test title 2",
        }
      }
    />
    <button
      className="button button--secondary"
      onClick={[MockFunction]}
    >
      Remove Expense
    </button>
    <ConfirmModal
      onRemove={[Function]}
      onRequestClose={[MockFunction]}
      postTitle="test title 2"
    />
  </div>
</div>
`;

exports[`should render error EditPost if uid is not macthed authorID 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Post
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <p
      className="message"
    >
      You don't allowed to edit this post
    </p>
    <Link
      className="button"
      replace={false}
      to="/dashboard"
    >
      Back to Dashboard
    </Link>
  </div>
</div>
`;

exports[`should render error if post doesn't exist 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Post
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <p
      className="message"
    >
      This post does not exist
    </p>
    <Link
      className="button"
      replace={false}
      to="/dashboard"
    >
      Back to Dashboard
    </Link>
  </div>
</div>
`;
